keps/sig-network/2943-networkpolicy-status/README.md

```diff
@@ -0,0 +1,462 @@
+
+	// Specification of the desired behavior for this NetworkPolicy.
+	// +optional
+	Spec NetworkPolicySpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+  // Status is the current state of the NetworkPolicy.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	// +optional
+	Status NetworkPolicyStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+```
+
+* Add the `NetworkPolicyStatus` struct as following:
+```go
+// NetworkPolicyStatus describe the current state of the NetworkPolicy.
+type NetworkPolicyStatus struct {
+  // Conditions holds an array of metav1.Condition that describe the state of the NetworkPolicy.
+ Conditions []metav1.Conditon
+}
+```
+being metav1.Condition:
+
+```go
+type Condition struct {
+	Type string `json:"type" protobuf:"bytes,1,opt,name=type"`
+	// status of the condition, one of True, False, Unknown.
+	Status ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status"`
+	// observedGeneration represents the .metadata.generation that the condition was set based upon.
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+	ObservedGeneration int64 `json:"observedGeneration,omitempty" protobuf:"varint,3,opt,name=observedGeneration"`
+	// lastTransitionTime is the last time the condition transitioned from one status to another.
+	// This should be when the underlying condition changed.  If that is not known, then using the time when the API field changed is acceptable.
+	// +required
+	LastTransitionTime Time `json:"lastTransitionTime" protobuf:"bytes,4,opt,name=lastTransitionTime"`
+	// reason contains a programmatic identifier indicating the reason for the condition's last transition.
+	// Producers of specific condition types may define expected values and meanings for this field,
+	// and whether the values are considered a guaranteed API.
+	// The value should be a CamelCase string.
+	// This field may not be empty.
+	Reason string `json:"reason" protobuf:"bytes,5,opt,name=reason"`
+	// message is a human readable message indicating details about the transition.
+	// This may be an empty string.
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=32768
+	Message string `json:"message" protobuf:"bytes,6,opt,name=message"`
+}
+```
+
+The above is an extraction from https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L1475-L1521
+
+```
+<<[UNRESOLVED default conditions ]>>
+Should Network Policy Status cover multiple network policy provider conditions? 
+If so, how can this be solved? Prefixing the Type field with the implementation
+name as `myprovider.com/PartialFailure?`
+
+Also, what kind of polatiry should we use in a default condition Type? Positive polarity 
+like "Accepted=bool" or negative polarity like PartialFailure=true?
+
+See the original discussion in https://github.com/kubernetes/enhancements/pull/2947#discussion_r796974848
```

Comment:
In Gateway API, we've settled on this, (based on my personal opinions as much as anything else):
- a small number (ideally one) of positive-polarity rollup conditions (`Accepted` would be the go here), but I could see some value in both `Accepted`, meaning the config is okay, and `Ready`, meaning the responsible controller has rolled the config out everywhere it needs to be. The idea for these is to provide a low-noise signal that either "everything is OK, no worries", or "something is wrong, you should check more". 
- If necessary, negative-polarity conditions that indicate the presence of a specific error (`IPNotAllowed`, or similar error conditions.)  The Node error conditions are the exemplar here, I think. Also note that providing a good Reason and Message on an `Accepted` condition may be sufficient, and that these negative-polarity conditions should probably start out namespaced to the owning controller until there's some agreement about what's useful.

Having a `Ready` Condition as well as an `Accepted` one allows a user to easily distinguish between "My config is busted" and "my config hasn't been deployed yet", without requiring all nodes to update the object. I'd guess that it should be relatively straightforward for the policy controller to have some mechanism for knowing which nodes have the config version, even if that's just polling.

For any required Conditions, I strongly recommend providing string constants for Reason values you want to be well known, and document the cases you expect them to be used as clearly as possible.

However, having these non-namespaced summary Conditions required will make it difficult to impossible for multiple controllers to coordinate. If multiple controllers are really required, *all* Conditions should be namespaced with the name of the owning controller, and the API definitions can only suggest Type suffixes. This can be done just using the existing Type field (since it's kind of expected that it may contain domain-prefixed strings).

I'd agree with a comment from @aojea that questioned the need to support multiple network policy providers, personally, but my knowledge in this area is pretty out-of-date, so take that with a whole pile of salt.

Reply:
> I'd agree with a comment from @aojea that questioned the need to support multiple network policy providers, personally, but my knowledge in this area is pretty out-of-date, so take that with a whole pile of salt.

+1 to this one, but again, I'm far from being an expert here.

My main question is - do we have to address that for Alpha? While I think we should resolve that for beta, given that alpha is disabled by default, we can probably start with a single provider support only and reevaluate for beta.
@thockin - thoughts?

Reply:
I think, IMO that in this case and as Tim stated before, we can deal with that on the implementation moment rather than right now. We are targeting alpha, so we can change as much as we want right? :)

Thanks folks for your thoughts on this!